### PR TITLE
[FIX] Consider fixed mods for terms, reading pepXML

### LINKED ISSUE #5538: "Non-fatal" error encountered when using IDFileConverter with dimethyl static modification
Hi, I'm using IDFileConverter 2.6.0 (Ubuntu 20.04) to convert output from comet 2019.01 and seeing errors like this:

```
<Non-fatal error while loading test.pep.xml': Cannot find N-terminal modification with mass 29.039124999999999.> occurred 40587 times
```

This is for samples that are dimethyl labeled, searched with a static mod for the light tag and a diff-mod for the heavy tag. I realize that this may be non-standard but I think it should work as it's a valid search strategy.

Here are the relevant modifications from the comet params file:

```
add_Nterm_peptide = 28.031300
add_K_lysine = 28.031300
variable_mod01 = 6.031817 nK 1 3 -1 0 0 0.0
variable_mod02 = 15.994915 M 0 3 -1 0 0 0.0 
```

To support this search strategy, I've also had to add an entry to `unimod.xml` which corresponds to the +6 mass above:

```xml
<umod:mod
    title="Custom Dimethyl:2H(4)13C(2)"
    full_name="Custom Dimethyl:2H(4)13C(2)"
    username_of_poster="Radu"
    group_of_poster=""
    date_time_posted="2021-01-06 12:00:00"
    date_time_modified="2021-01-06 12:00:00"
    approved="0"
    record_id="99901"
>
    <umod:specificity hidden="1" site="N-term" position="Protein N-term" classification="Isotopic label" spec_group="4">
        <umod:misc_notes>When dimethyl labelling is pre-digest</umod:misc_notes>
    </umod:specificity>
    <umod:specificity hidden="1" site="R" position="Anywhere" classification="Isotopic label" spec_group="3"/>
    <umod:specificity hidden="1" site="K" position="Anywhere" classification="Isotopic label" spec_group="1"/>
    <umod:specificity hidden="1" site="N-term" position="Any N-term" classification="Isotopic label" spec_group="2"/>
    <umod:delta mono_mass="6.031817" avge_mass="6.009956776" composition="13C(2) 2H(4) C(-2) H(-4)">
        <umod:element symbol="13C" number="2"/>
        <umod:element symbol="2H" number="4"/>
        <umod:element symbol="C" number="-2"/>
        <umod:element symbol="H" number="-4"/>
    </umod:delta>
</umod:mod>
```

And here are the lines of the .pep.xml comet output where modifications are specified:

```xml
<aminoacid_modification aminoacid="K" massdiff="6.031817" mass="162.158080" variable="Y" binary="Y" description="Custom Dimethyl:2H(4)13C(2) (K)" />
<aminoacid_modification aminoacid="M" massdiff="15.994915" mass="147.035400" variable="Y" description="Oxidation (M)" />
<terminal_modification terminus="N" massdiff="6.031817" mass="35.070942" variable="Y" protein_terminus="N" description="Custom Dimethyl:2H(4)13C(2) (N-term)" />
<terminal_modification terminus="N" massdiff="28.031300" mass="29.039125" variable="N" protein_terminus="N" description="Dimethyl (N-term)" />
<aminoacid_modification aminoacid="C" massdiff="57.021464" mass="160.030649" variable="N" description="Carbamidomethyl (C)" />
<aminoacid_modification aminoacid="K" massdiff="28.031300" mass="156.126263" variable="N" description="Dimethyl (K)" />
```

Note that I've added descriptions to each to match `unimod.xml`. Then, running `IDFileConverter` as follows:

```
IDFileConverter -in "test.pep.xml" -out "test.idXML" -threads $(nproc)
Progress of 'Loading...':Non-fatal error while loading 'test.pep.xml': Cannot find N-terminal modification with mass 29.039124999999999.
-- done [took 6.91 s (CPU), 6.91 s (Wall)] -- 
Progress of 'Storing...':
Omitted writing of 119 peptide identifications due to empty hits.
-- done [took 2.08 s (CPU), 2.14 s (Wall)] -- 
IDFileConverter took 9.14 s (wall), 9.07 s (CPU), 0.39 s (system), 8.68 s (user); Peak Memory Usage: 114 MB.
<Non-fatal error while loading 'test.pep.xml': Cannot find N-terminal modification with mass 29.039124999999999.> occurred 40587 times
```

If we take the same .pep.xml file and "convert" (this is the same search) both light and heavy dimethyl mods to variable mods by swapping out the +6 mod with +34 (and now pointing to the non-custom mod), and making the +28 mod variable instead like so:

```xml
<aminoacid_modification aminoacid="K" massdiff="34.06312" mass="162.158080" variable="Y" binary="Y" description="Dimethyl:2H(4)13C(2) (K)" />
<aminoacid_modification aminoacid="M" massdiff="15.994915" mass="147.035400" variable="Y" description="Oxidation (M)" />
<terminal_modification terminus="N" massdiff="34.06312" mass="35.070942" variable="Y" protein_terminus="N" description="Dimethyl:2H(4)13C(2) (N-term)" />
<terminal_modification terminus="N" massdiff="28.031300" mass="29.039125" variable="Y" protein_terminus="N" description="Dimethyl (N-term)" />
<aminoacid_modification aminoacid="C" massdiff="57.021464" mass="160.030649" variable="N" description="Carbamidomethyl (C)" />
<aminoacid_modification aminoacid="K" massdiff="28.031300" mass="156.126263" variable="Y" description="Dimethyl (K)" />
```

we get no errors:

```
$ IDFileConverter -in "test.pep.xml" -out "test2.idXML" -threads $(nproc)
Progress of 'Loading...':
-- done [took 5.93 s (CPU), 5.94 s (Wall)] -- 
Progress of 'Storing...':
Omitted writing of 119 peptide identifications due to empty hits.
-- done [took 2.13 s (CPU), 2.18 s (Wall)] -- 
IDFileConverter took 8.23 s (wall), 8.17 s (CPU), 0.43 s (system), 7.74 s (user); Peak Memory Usage: 114 MB.
```

Now if we compare the .idXML outputs, after removing the "Custom " prefix we gave our dimethyl mod earlier, we see that there is no difference, other than the static mods being listed as static now:

```bash
$ sed -i 's/Custom //g' test.idXML
$ diff test.idXML test2.idXML
5d4
<               <FixedModification name="Dimethyl (N-term)" />
7d5
<               <FixedModification name="Dimethyl (K)" />
10a9,10
>               <VariableModification name="Dimethyl (N-term)" />
>               <VariableModification name="Dimethyl (K)" />
```

So, being new to OpenMS, I don't know much about what is going on internally, but it seems to me that there might be an issue with how N-terminal static mods are being handled, albeit one that does not produce any difference in output, unless you consider the "non-fatal" error.

Happy to help dissect this further - please let me know if there's any other information I can provide. Also, I should mention that I'm aware that there are other ways to search dimethyl labeled samples, and that this might be non-standard, but for now I would prefer not to deviate from established protocol.

### PR DESCRIPTION
**Pull request by jpfeuffer**
# Description

Maybe fixes #5538

Also allows reading unknown scores from pepXML as requested on Gitter.

### CODE REVIEW

File: src/openms/source/FORMAT/PepXMLFile.cpp [hunk 1]
```diff
@@ -958,9 +959,15 @@ namespace OpenMS
     }
   }
 
+  void PepXMLFile::setParseUnknownScores(bool parse_unknown_scores)
+  {
+    this->parse_unknown_scores_ = parse_unknown_scores_;
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/PepXMLFile.cpp:964]: Remove last underscore … wrong this way

File: src/openms/source/FORMAT/PepXMLFile.cpp [hunk 2]
```diff
@@ -1320,11 +1345,9 @@ namespace OpenMS
       {
         current_peptide_.setMetaValue("pepxml_spectrum_name", native_spectrum_name_);
       }
-      if (search_engine_ == "Comet")
-      {
-        current_peptide_.setMetaValue("spectrum_reference", native_spectrum_name_);
-        //TODO: we really need something uniform here, like scan number - and not in metainfointerface
-      }
+      //TODO: we really need something uniform here, like scan number - and not in metainfointerface
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/PepXMLFile.cpp:1348]: not sure if scan number is better than e.g. native id once we also deal with ion mobility frames etc.

File: src/openms/source/FORMAT/PepXMLFile.cpp [hunk 3]
```diff
@@ -1467,7 +1490,15 @@ namespace OpenMS
             break; // only one modification should match, so we can stop the loop here
           }
         }
-        //TODO why only look in variable mods?
+        for (const AminoAcidModification& it : fixed_modifications_)
+        {
+          if ((fabs(mod_nterm_mass - it.getMass()) < mod_tol_) && it.getTerminus() == "n")
+          {
+            current_modifications_.emplace_back(it.getRegisteredMod(), 42); // position not needed for terminus
```
[the remark this answers is not in the document]
Comment (jpfeuffer) [reply src/openms/source/FORMAT/PepXMLFile.cpp:1497]: It is unused as the comment implies. You can overload the function if you want.

File: src/openms/source/FORMAT/PepXMLFile.cpp (hunk 2 above)
Comment (jpfeuffer) [reply src/openms/source/FORMAT/PepXMLFile.cpp:1348]: Just copied. But the main gist is true. It should be a fixed field.

File: src/openms/source/FORMAT/PepXMLFile.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/FORMAT/PepXMLFile.cpp:1497]: maybe -1?

**Comment by jpfeuffer** [comment]
@timosachsenberg What do we do about the failed tests? I really want to parse spectrum_reference always (not only for Comet), even if it contains non-sense (which it doesn't so much nowadays). But in our test cases, there is a lot of non-sense.

File: src/openms/source/FORMAT/PepXMLFile.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpfeuffer) [reply src/openms/source/FORMAT/PepXMLFile.cpp:1497]: I think it is Size no?

[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/FORMAT/PepXMLFile.cpp:1497]: I think you can still overload it with -1

**Comment by timosachsenberg** [comment]
what does nonsense mean in that case?

**Comment by jpfeuffer** [comment]
see tests: spectrum_reference="PepXMLFile_test.00002.00002.3"

**Comment by jpfeuffer** [comment]
It's basically the PeptideProphet referencing

**Comment by timosachsenberg** [comment]
But there we can't link back to the spectra file? We could create a new identifier from that e.g.
 convert this id to "scan=2"

https://github.com/HUPO-PSI/mzTab/blob/master/specification_document-releases/2_0-Metabolomics-Release/mzTab_format_specification_2_0-M_release.adoc#51-use-of-identifiers-for-input-spectra-to-a-search
e.g. MS:1000776

**Comment by jpfeuffer** [comment]
I guess we could, as long as it is unique. Yeah but I don't know about the standard.

**Comment by timosachsenberg** [comment]
I mean it is a hack anyway.
Use the poper native ID if we have one. or if it is in this format convert it to "scan=..."

**Comment by jpfeuffer** [comment]
But when is it a proper native ID?

**Comment by jpfeuffer** [comment]
The last entry in your table actually allows arbitrary strings.

**Comment by timosachsenberg** [comment]
I think @oliveralka has written some code to detect and parse them

**Comment by oliveralka** [comment]
Unfortunately, I think the use-case was totally different. 

We wanted to extractTheScanNumber based on the `native_id` and `native_id_type_accession`.
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/METADATA/SpectrumLookup.cpp#L259

**Review by timosachsenberg — APPROVED** [review]
looks good. any comment you want to add to the PR regarding the gitpod change?

**Comment by jpfeuffer** [comment]
Just a newer clangd. And such that it is found from the beginning.

**Comment by jpfeuffer** [comment]
we need to disable bg indexing bc of crashes in some files.

**Comment by timosachsenberg** [comment]
rebuild jenkins

**Pull request merged**